chapters/classes.tex

```diff
@@ -276,36 +276,111 @@ \subsubsection{Prefix Rules}\label{prefix-rules}
+  can only change its value at event
+  instants (see \cref{events-and-synchronization}).
+\item
+  A \emph{continuous-time} variable \lstinline!vn! may have a non-vanishing time
+  derivative (\lstinline!der(vn)<>0! possible) and may also
+  change its value discontinuously at any time during transient analysis
+  (see \cref{events-and-synchronization}). If there are any discontinuities the variable is
+  not differentiable.
+\end{itemize}
+
+\begin{nonnormative}
+For a parameter in a valid model, presence of \lstinline!Evaluate! (\cref{modelica:Evaluate}) makes it possible to tell immediately whether it is a structural or normal parameter, since \lstinline!Evaluate = false! makes it a normal parameter by definition, and \lstinline!Evaluate = true! is only allowed if the parameter is structural.
+\end{nonnormative}
+
+By the acyclic binding rule in \cref{acyclic-bindings-of-constants-and-parameters}, it follows that the value of a constant or structural parameter to be used in simplifications is possible to obtain by evaluation of a structural expression where values are available for all component subexpressions.
 
+\begin{example}
+A particularly demanding aspect of this evaluation is the potential presence of external functions.
+Hence, if it is known that a parameter won't be used by a structural expression, a user can make it clear that the external function is not meant to be evaluated during translation by using \lstinline!Evaluate = false!:
+\begin{lstlisting}[language=modelica]
+import length = Modelica.Utilities.Strings.length; /* Pure external function */
+parameter Integer n = length("Hello");             /* Structural parameter */
+parameter Integer p = length("Hello")
+  annotation(Evaluate = false);                    /* Normal parameter */
+parameter Boolean b = false;
+
+/* Variability analysis might cause evaluation of n: */
+parameter Real x = if b and n > 3 then x else 0;
+
+/* Variability analysis cannot cause evaluation of p: */
+parameter Real y = if b and p > 3 then y else 0;
+\end{lstlisting}
+\end{example}
+
+\begin{example}
+Normal and structural parameters.
+\begin{lstlisting}[language=modelica]
 model ABCD
-  parameter Real A[n, n];
+  parameter Real A[n, n]; // Requires that n is a structural parameter.
   parameter Integer n = size(A, 1);
 end ABCD;
 
 final ABCD a;
-// Illegal since cyclic dependencies between size(a.A,1) and a.n
+// Illegal since cyclic dependencies between size(a.A, 1) and a.n makes n a normal parameter.
 
 ABCD b(redeclare Real A[2, 2] = [1, 2; 3, 4]);
-// Legal since size of A is no longer dependent on n.
+// Legal since size of A is no longer dependent on n, making n a structural parameter.
 
 ABCD c(n = 2); // Legal since n is no longer dependent on the size of A.
 
-parameter Real r = 2 * sin(r); // Illegal, since r = 2 * sin(r) is cyclic
+parameter Real r = 2 * sin(r); // Normal parameter, since r = 2 * sin(r) is cyclic.
```

Comment:
As before, I don't see why we want to allow these cases.

Reply:
Because it avoids the complicated corner cases of the old acyclic binding rule if we exclude normal parameters (including the non-fixed ones) from the analysis.  The importance of this sort of rule should be to make sure constants and structural parameters can be evaluated during translation – systems of equations in the initialization problem involving normal parameters is something we should consider normal in Modelica.

Reply:
Systems of equations involving parameters may be normal, but writing them in terms of parameter bindings is error-prone and if pushed too far will lead to convoluted binding equations. If I see:
`  parameter Real r=2*sin(r); `
I would not be sure what was intended. If I see:
```
  parameter Real r(fixed=false);
initial equation
  r/2=sin(r);
```
I'm sure that it is intended to solve r from an initial equation. Basically one major reason we introduce initial equations (compared to the previous work-arounds) was to increase readability.

Reply:
I agree that the use of `initial equation` is nice in this case.  However, we're coming from the situation described in #2742 where binding equations could still end up in systems of equations due to interactions with non-fixed parameters (that don't use binding equations, and thereby circumvents the current acyclic binding rule).

I'd be fine with suggesting a warning to be issued for the case of a cycle of normal parameters where all have declaration equations.

Reply:
I agree with Hans. Allowing cyclic bindings for normal parameters seem unnecessary and out of scope for the purposes of this MCP.

Reply:
The warning could also be combined with – or replaced by – making it an error if there is a cycle involving only parameter declaration equations (no holes in the cycle corresponding to non-fixed parameters' initial equations).

Reply:
A warning does not clarify to me how to interpret that expression, as Hans said. One may choose to interpret it as ``r = 2 * sin(r.start)``.

Reply:
This has now been resolved by not allowing the cycles, I believe.

Reply:
Yes, the current form of the example is essentially the same as before:
```
/* All of the following are illegal: */
parameter Real r = 2 * sin(r); // Depends directly on r.
…
```
Resolving.